gdnative-core/src/export/method.rs

```diff
@@ -284,16 +287,303 @@ impl<'a> Varargs<'a> {
     type Item = &'a Variant;
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        self.iter.next().copied()
+        let ret = self.args.get(self.idx).copied();
+        if ret.is_some() {
+            self.idx += 1;
+        }
+        ret
+    }
+}
+
+/// All possible error types for convert from Varargs.
+#[derive(Debug)]
+pub enum VarargsError {
+    ArgumentTypeError(ArgumentTypeError),
+    ArgumentLengthError(ArgumentLengthError),
+}
+
+impl std::error::Error for VarargsError {}
+impl std::fmt::Display for VarargsError {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self {
+            VarargsError::ArgumentTypeError(e) => e.fmt(f),
+            VarargsError::ArgumentLengthError(e) => e.fmt(f),
+        }
+    }
+}
+
+impl From<ArgumentTypeError> for VarargsError {
+    #[inline]
+    fn from(value: ArgumentTypeError) -> Self {
+        Self::ArgumentTypeError(value)
+    }
+}
+
+impl From<ArgumentLengthError> for VarargsError {
+    #[inline]
+    fn from(value: ArgumentLengthError) -> Self {
+        Self::ArgumentLengthError(value)
+    }
+}
+
+/// Error to incorrect type of argument.
+/// Displays a message containing the position of the argument and cause of the failure to convert.
+#[derive(Debug)]
+pub struct ArgumentTypeError {
+    index: usize,
+    nested_error: FromVariantError,
+}
+
+impl ArgumentTypeError {
+    /// Create a new error with the argument position and `FromVariantError`.
+    #[inline]
+    #[must_use]
+    pub fn new(index: usize, nested_error: FromVariantError) -> Self {
+        Self {
+            index,
+            nested_error,
+        }
+    }
+
+    /// Returns an ordinal number representation.
+    #[inline]
+    #[must_use]
+    fn ordinal(&self) -> String {
+        match self.index + 1 {
+            1 => "1st".to_owned(),
+            2 => "2nd".to_owned(),
+            3 => "3rd".to_owned(),
+            i @ 4.. => format!("{i}th"),
+            _ => "unknown".to_owned(),
+        }
+    }
+
+    /// Get the argument type error's index.
+    #[inline]
+    #[must_use]
+    pub fn index(&self) -> usize {
+        self.index
+    }
+
+    /// Get a reference to the argument type error's nested error.
+    #[inline]
+    #[must_use]
+    pub fn nested_error(&self) -> &FromVariantError {
+        &self.nested_error
+    }
+}
+
+impl std::error::Error for ArgumentTypeError {}
+impl std::fmt::Display for ArgumentTypeError {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Incorrect type of {} argument, cause: {}",
+            self.ordinal(),
+            self.nested_error,
+        )
+    }
+}
+
+/// Error to argument lengths do not match.
+/// Display a message containing the length of arguments passed and the expected range of lengths.
+#[derive(Debug)]
+pub struct ArgumentLengthError {
+    passed: usize,
+    expected: (Bound<usize>, Bound<usize>),
+}
```

Comment:
Errors in godot-rust typically don't have rich APIs, e.g. [`FromVariantError`](https://godot-rust.github.io/docs/gdnative/core_types/enum.FromVariantError.html) is simply an enum without methods.

[`FromVariantError::InvalidLength`](https://godot-rust.github.io/docs/gdnative/core_types/enum.FromVariantError.html#variant.InvalidLength) in particular is a good inspiration for this implementation:
```rs
    ...
    /// Length of the collection is different from the expected one.
    InvalidLength { 
        len: usize, 
        expected: usize 
    },
    ...
```

I think we could simply make the fields public here, and use a more similar naming:
```rs
#[derive(Debug)]
pub struct InvalidArgumentLength { // or *Count?
    pub len: usize,
    pub expected_min: Bound<usize>,
    pub expected_max: Bound<usize>,
}
```
Then, we would not need the 3 methods `passed()`, `expected_min()` and `expected_max()` -- it's anyway not clear if returning `usize::MIN`/`usize::MAX` is intuitive for the user -- and with `Bounds`, the user can work with a higher-level API.

In 99% of the cases, a user would not even need to access these values programmatically, but simply print the error message.

Reply:
**TLDR:** let's keep rarely-used APIs as simple as possible, with only as much code as necessary. 
We can always add more features in the future if there is actual demand.

Reply:
> I think we could simply make the fields public here,

In that case, the field becomes mutable.

> Then, we would not need the 3 methods passed(), expected_min() and expected_max() -- it's anyway not clear if returning usize::MIN/usize::MAX is intuitive for the user -- and with Bounds, the user can work with a higher-level API.

Which is the better choice? 🤔 

> let's keep rarely-used APIs as simple as possible, with only as much code as necessary.

If it is an API that users can implement on their own, there is no problem without it, but since it is impossible to implement, this is a major limitation for users.

Also, just because there are more lines of code does not mean that they are more complex.

> We can always add more features in the future if there is actual demand.

Will you investigate demand later? I think that would be more labor intensive.

Reply:
> In that case, the field becomes mutable.

Yes, just like the `VariantError` enum.

---

> If it is an API that users can implement on their own, there is no problem without it, but since it is impossible to implement, this is major limitation for users.

What do you mean here? We expose all the relevant info with `len`, `expected_min` and `expected_max` being public. We can also keep `Display` for convenience. The user can do everything, no?

---

> > We can always add more features in the future if there is actual demand.
>
> Will you investigate demand later? I think that would be more labor intensive.

First, a user expects similar functionality to have similar APIs. `VariantError` variants  is an mutable, publicly accessible fields (implied by being an enum).
```rs
    /// Length of the collection is different from the expected one.
    InvalidLength { len: usize, expected: usize },
    /// Invalid enum representation.
    InvalidEnumRepr {
        expected: VariantEnumRepr,
        error: Box<FromVariantError>,
    },
```
Now `VarargsError` is completely different: it uses tuple-like enumerators instead of struct-like ones, and its external structs are encapsulating their fields. This inconsistency exists despite both APIs serving an identical purpose.

Regarding the demand part of your question -- I'm a big fan of "keep it simple". No one ever demanded highly encapsulated error APIs. Most people don't even care about the error -- they `unwrap()` and if there's a panic, it's a bug to be fixed. At most, people may output the error message. I guess we could even use `anyhow::Error` to throw a string inside, and most users would still be happy. But we're already providing much more than that -- the user can differentiate _precisely_ which error happened and react to it in a very specific way, if necessary. I don't think anything _beyond_ this (such as `From`, or separate struct types, or encapsulation) adds real value to the library.

If I am wrong in my opinion, people have the option to use the issue tracker. I personally believe it's much less labor-intensive to start with little code and expand when needed, rather than put a lot of up-front effort for a corner-case API and make assumptions how this going to benefit everyone.

Reply:
Serious question: why don't we go with this? It's like `FromVariantError` does it.
```rs
pub enum VarargsError {
    ArgumentTypeError {
        index: usize,
        error: FromVariantError, // or Box
    },
    ArgumentLengthError {
        len: usize,
        expected_min: Bound<usize>,
        expected_max: Bound<usize>,
    },
}
```